samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java

```diff
@@ -0,0 +1,212 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.table.remote.couchbase;
+
+import com.couchbase.client.java.Bucket;
+import com.couchbase.client.java.Cluster;
+import com.couchbase.client.java.CouchbaseCluster;
+import com.couchbase.client.java.env.CouchbaseEnvironment;
+import com.couchbase.client.java.env.DefaultCouchbaseEnvironment;
+import com.couchbase.client.java.util.retry.RetryWhenFunction;
+import com.google.common.collect.ImmutableList;
+import java.io.Serializable;
+import java.security.KeyStore;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import org.apache.samza.context.Context;
+import org.apache.samza.operators.functions.ClosableFunction;
+import org.apache.samza.operators.functions.InitableFunction;
+import org.apache.samza.serializers.Serde;
+
+
+public abstract class BaseCouchbaseTableFunction<V> implements InitableFunction, ClosableFunction, Serializable {
+
+  // Clients
+  protected transient CouchbaseEnvironment env;
+  protected transient Cluster cluster;
+  protected transient Bucket bucket;
+
+  // Function Settings
+  protected Class<V> valueClass = null;
+  protected Serde<V> valueSerde = null;
+  protected long timeout = 0L;
+  protected TimeUnit timeUnit = null;
+  protected int ttl = 0; // default value 0 means no ttl, data will be stored forever
+
+  // Cluster Settings
+  protected List<String> clusterNodes = null;
+  protected String username = null;
+  protected String password = null;
+  protected String bucketName = null;
+
+  // Environment Settings
+  protected boolean sslEnabled = false;
+  protected boolean certAuthEnabled = false;
+  protected String sslKeystoreFile = null;
+  protected String sslKeystorePassword = null;
+  protected String sslTruststoreFile = null;
+  protected String sslTruststorePassword = null;
+  protected int bootstrapCarrierDirectPort = -1;
+  protected int bootstrapCarrierSslPort = -1;
+  protected int bootstrapHttpDirectPort = -1;
+  protected int bootstrapHttpSslPort = -1;
+
+  public BaseCouchbaseTableFunction() {
+  }
+
+  //TODO maybe we can create a builder class to create both read and write functions for the same bucket so that users don't need to type in the same things twice
+  public BaseCouchbaseTableFunction(Class<V> valueClass) {
```

Comment:
Would it make sense to put mandatory parameters such as cluster, bucket name, etc. in the ctor?

Reply:
Yes, that makes sense.